### DISCUSSION TITLE
UnknownCryptoError when encrypting

### DISCUSSION THREAD

**Discussion by Jytesh** [orion-rs/orion#204] [Q&A] [answered]
```rs
use orion::aead as crypto;

pub fn enrypt_string(string: String, password: String) -> Result<Vec<u8>, orion::errors::UnknownCryptoError>{
    println!("{:?}", password.as_bytes());
    let passphrase = crypto::SecretKey::from_slice(password.as_bytes()).expect("Failed to encrypt password");
    println!("{:?} {:?}", passphrase, string.as_bytes());
    let ciphertext = crypto::seal(&passphrase, &string.as_bytes()).expect("Failed to encrypt text");
    Ok(ciphertext)
}
```

Panics at
```
normalpassword
TESTING CRYPTO
[110, 111, 114, 109, 97, 108, 112, 97, 115, 115, 119, 111, 114, 100]
SecretKey {***OMITTED***} [84, 69, 83, 84, 73, 78, 71, 32, 67, 82, 89, 80, 84, 79]
thread 'main' panicked at 'Failed to encrypt text: UnknownCryptoError', src/crypto.rs:7:68
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by Jytesh** [comment]
I changed the secretkey method to default() and it works now, but I want a user inputted password

**Comment by brycx** [comment, marked as answer]
Hi!

If you take a peek at the [errors documentation](https://docs.rs/orion/0.16.0/orion/aead/index.html#errors), you'll see that the first one says that the `secret_key` must be 32 bytes. The password you use here is however only 14 bytes. You can use the `.len()` function on a `SecretKey` instance to check if it is upheld. `default()` randomly generates 32 bytes, which is why it works.

In practice, please note that using user-passwords directly to encrypt data is not secure. You can use [`orion::kdf`](https://docs.rs/orion/0.16.0/orion/kdf/index.html) to "stretch" the password before using it for encryption. There is also an example in that module for this exact use-case.

Don't hesitate to let us know if there's anything else!

**Reply by Jytesh** [reply]
Thank you I will try I am fairly new to crypto and this crate is really cool :D

**Reply by Jytesh** [reply]
If i do stretch the password then can I use the same user given password to decrypt the ciphertext?

**Reply by brycx** [reply]
If you use the derived key (the result of stretching the password), you'll have to use the same key to decrypt. This means that, if you have `encrypt_string` and `decrypt_string`, that take in the non-stretched password, like above, then you'll have to stretch it in `decrypt_string` as well. If you use the same cost parameter for stretching and the same salt, then the same stretched key will be created.

**Reply by Jytesh** [reply]
Alright thank you!

**Reply by brycx** [reply]
No problem!

**Reply by Jytesh** [reply]
How would I use the same salt across multiple runs? I encrypt some text and store it to file and later i want to decrypt it with the user given password?

**Reply by Jytesh** [reply]
```

fn stretch_passphrase(passphrase: String) -> crypto::SecretKey {
    let user_password = kdf::Password::from_slice(passphrase.as_bytes()).unwrap();
    let salt = kdf::Salt::default();

    let derived_key = kdf::derive_key(&user_password, &salt, 3, 1<<16, 32).unwrap();
    return derived_key;
}
```
Current doing this but the default salt is randomly generated so the stretched password isn't the same

**Reply by brycx** [reply]
If you have a file, the common approach is to randomly generate the salt for a given file, write the salt to the file as the first n bytes, then write the encrypted data afterwards. Then, when you want to decrypt, you read the salt from the encrypted file and use that to re-create the derived key. Afterwards, you can decrypt the rest of the file.

**Reply by Jytesh** [reply]
Okay I will try that

**Reply by Jytesh** [reply]
works! I have successfully encrypted and decrypted! :heart:

**Reply by Jytesh** [reply]
https://github.com/Jytesh/file-crypto-cli :)
